src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1996,52 +1958,66 @@ private TSet ComputeStartSet()
                 case SymbolicRegexNodeKind.BOLAnchor:
                 case SymbolicRegexNodeKind.CaptureStart:
                 case SymbolicRegexNodeKind.CaptureEnd:
-                    return _builder._solver.Empty;
+                    return builder._solver.Empty;
 
                 case SymbolicRegexNodeKind.Singleton:
                     Debug.Assert(_set is not null);
                     return _set;
 
                 case SymbolicRegexNodeKind.Loop:
                     Debug.Assert(_left is not null);
-                    return _left._startSet;
+                    return _left.GetStartSet(builder);
 
                 case SymbolicRegexNodeKind.Concat:
                     {
                         Debug.Assert(_left is not null && _right is not null);
-                        TSet startSet = _left.CanBeNullable ? _builder._solver.Or(_left._startSet, _right._startSet) : _left._startSet;
+                        TSet startSet = _left.CanBeNullable ? builder._solver.Or(_left.GetStartSet(builder), _right.GetStartSet(builder)) : _left.GetStartSet(builder);
                         return startSet;
                     }
 
                 case SymbolicRegexNodeKind.Alternate:
                     {
                         Debug.Assert(_left is not null && _right is not null);
-                        return _builder._solver.Or(_left._startSet, _right._startSet);
+                        return builder._solver.Or(_left.GetStartSet(builder), _right.GetStartSet(builder));
                     }
 
                 case SymbolicRegexNodeKind.DisableBacktrackingSimulation:
                 case SymbolicRegexNodeKind.Effect:
                     Debug.Assert(_left is not null);
-                    return _left._startSet;
+                    return _left.GetStartSet(builder);
 
                 default:
-                    Debug.Fail($"{nameof(ComputeStartSet)}:{_kind}");
-                    return _builder._solver.Full;
+                    Debug.Fail($"{nameof(GetStartSet)}:{_kind}");
+                    return builder._solver.Full;
             }
         }
 
         /// <summary>
         /// Replace anchors that are infeasible by [] wrt the given previous character kind and what continuation is possible.
         /// </summary>
+        /// <param name="builder">the builder that owns this node</param>
         /// <param name="prevKind">previous character kind</param>
-        /// <param name="contWithWL">if true the continuation can start with wordletter or stop</param>
-        /// <param name="contWithNWL">if true the continuation can start with nonwordletter or stop</param>
-        internal SymbolicRegexNode<TSet> PruneAnchors(uint prevKind, bool contWithWL, bool contWithNWL)
+        internal SymbolicRegexNode<TSet> PruneAnchors(SymbolicRegexBuilder<TSet> builder, uint prevKind)
```

Comment:
Can you add a comment about why this is necessary?

Reply:
This function helps avoid doing spurious work in some situations where the pattern's language is empty due to interactions between anchors and the rest of the pattern, but where the pattern isn't actually the "nothing" node. One example would be `a*\ba`, which would be simplified to `[]` when the `prevKind` is a word letter. It's essentially supporting the dead end state detection notice non-trivial cases.

I'll add a comment.